### PR TITLE
refactor: change to top level project locator.

### PR DESCRIPTION
**Pull request by hugorut**
This change hoists the `hcl.ProjectLocator` to the `providers.Detect` function. This is required before we tackle Terragrunt auto-detect support. This change means:

* `HCLProvider`s now only have one associated Parser - a `HCLProvider` is 1-1 linked with a project.
* `providers.Detect` functionality happens before we kick off the concurrent project run functionality. We "detect" all the possible projects and then send these to the job queue.

### CODE REVIEW

**Comment by hugorut** [comment]
test failures look unrelated

File: internal/hcl/block.go [hunk 1]
```diff
@@ -113,7 +113,7 @@ func (b referencedBlocks) Less(i, j int) bool {
 	return false
 }
 
-// ModuleBlocks is a wrapper around SortedByCaller that selects just Modules to be sorted.
+// ModuleBlocks is a wrapper around SortedByCaller that selects just Module to be sorted.
```
Comment (aliscott) [inline internal/hcl/block.go:116]: Looks like a find and replace issue

File: internal/hcl/modules/manifest.go [hunk 2]
```diff
@@ -17,7 +17,7 @@ type Manifest struct {
 
 	Path    string            `json:"Path"`
 	Version string            `json:"Version"`
-	Modules []*ManifestModule `json:"Modules"`
+	Modules []*ManifestModule `json:"Module"`
```
Comment (aliscott) [inline internal/hcl/modules/manifest.go:20]: Looks like a find and replace issue

File: internal/hcl/modules/registry.go [hunk 3]
```diff
@@ -359,7 +360,7 @@ func findLatestMatchingVersion(versions []string, constraints string) (string, e
 // This does not mean that the module address is a registry module, it could still be a remote module.
 // To work that out we need to try looking up the module using the `lookupModule` function
 func normalizeRegistrySource(moduleAddr string) (string, error) {
-	// Modules are in the format (registry)/namspace/module/target
+	// Module are in the format (registry)/namspace/module/target
```
Comment (aliscott) [inline internal/hcl/modules/registry.go:363]: Looks like a find and replace issue

File: internal/hcl/parser.go [hunk 4]
```diff
@@ -460,7 +343,7 @@ func (p *Parser) YAML() string {
 
 // ParseDirectory parses all the terraform files in the initialPath into Blocks and then passes them to an Evaluator
 // to fill these Blocks with additional Context information. Parser does not parse any blocks outside the root Module.
-// It instead leaves ModuleLoader to fetch these Modules on demand. See ModuleLoader.Load for more information.
+// It instead leaves ModuleLoader to fetch these Module on demand. See ModuleLoader.Load for more information.
```
Comment (aliscott) [inline internal/hcl/parser.go:346]: Looks like a find and replace issue

**Review by aliscott — APPROVED** [review]

**Pull request merged**